### PR TITLE
Feature/p icontroller

### PR DESCRIPTION
**Pull request by NicolaiFransen**
I am not completely happy about the torque controller, so please leave some nice comments if you have a good idea¿?
It should be like this:
PI with anit-windup

### CODE REVIEW

File: Software/Development/SemesterProject/App/PIController.c [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    PIcontroller->integralPart = 0;
+    PIcontroller->previousOutput = 0;
+    PIcontroller->previousLimitedOutput = 0;
+}
+
+void initPIControllers(void)
+{
+    PIObject_Constructor(&PIControllerList.IdController, 10, 10, 10);
+    PIObject_Constructor(&PIControllerList.IdController, 10, 10, 10);
```
Comment (AitorTeran) [inline Software/Development/SemesterProject/App/PIController.c]: Should one of this be IqController?

File: Software/Development/SemesterProject/App/PIController.c [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+/*
+ * PI controller for the inner loops.
+ */
+float PiCurrentCalculation(PIobject *PIcontroller, float measureredValue)
+{
+    int KP = PIcontroller->KP;
+    int KI = PIcontroller->KI;
+
+    float error, PiOutput;
```
Comment (AitorTeran) [inline Software/Development/SemesterProject/App/PIController.c]: make it PIoutput? same as PIcontroller

File: Software/Development/SemesterProject/App/PIController.c [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    // Calculating the integral part from the error
+    PIcontroller->integralPart = PIcontroller->integralPart + error;
+
+    // Calculating the difference between the wanted output and the actual output
+    // To be used in the anti-windup of the integrator
+    outputFeedback = PIcontroller->previousOutput - PIcontroller->previousLimitedOutput;
+
+    // Calculating the controller output
+    PiOutput = KP * error +
+              (KI * PIcontroller->integralPart - outputFeedback * 1/KP);
```
Comment (AitorTeran) [inline Software/Development/SemesterProject/App/PIController.c]: why "* 1 /"? why not taking the 1 out?

**Review by AitorTeran — COMMENTED** [review]
I've never implemented a PI in code but it looks good to me.
Good job!

File: Software/Development/SemesterProject/App/PIController.c (hunk 1 above)
Comment (NicolaiFransen) [reply Software/Development/SemesterProject/App/PIController.c]: Oh yes...

**Comment by NicolaiFransen** [comment]
K-values and saturation limits in Constant.h will have to be updated at some point

File: Software/Development/SemesterProject/App/PIController.c [hunk 4]
```diff
@@ -0,0 +1,141 @@
+/*
+ * PIController.c
+ *
+ *  Created on: 2. apr. 2019
+ *      Author: Nicolai Fransen
+ *
+ *  This file includes the PI controllers for the control system.
+ *  Two different PI functions have been created because the outer loop
```
Comment (murgui) [inline Software/Development/SemesterProject/App/PIController.c]: I think the 'two different functions' sentence is talking about deprecated software.

Maybe a bit more explanation of the general use and a list of the features that the code can do would be nice.

File: Software/Development/SemesterProject/App/PIController.c [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+void initPIControllers(void)
+{
+    PIObject_Constructor(&PIControllerList.IdController, KP_ID, KI_ID, CURRENT_LIMIT, 0);
+    PIObject_Constructor(&PIControllerList.IqController, KP_IQ, KI_IQ, CURRENT_LIMIT, 0);
+    PIObject_Constructor(&PIControllerList.SpeedController, KP_SPEED, KI_SPEED, SPEED_LIMIT, 1);
+}
+
+/*
+ * PI controller calculations.
+ * Depending on the value of the windup falg, this part will be used or not.
```
Comment (murgui) [inline Software/Development/SemesterProject/App/PIController.c]: flag typo

File: Software/Development/SemesterProject/App/Include/PIController.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+//
+#include "Constants.h"
+#include "DSP28x_Project.h"
+
+
+typedef struct PIobjectTag
+{
+    float KP;
+    float KI;
+    float integralPart;
```
Comment (murgui) [inline Software/Development/SemesterProject/App/Include/PIController.h]: I dont want to be whiny, but I think integralPart is not a name that is telling what is the variable for. Maybe you can come up with a better name¿¿ personally, I'd rather see a 4 word variable name ¯\_(ツ)_/¯

File: Software/Development/SemesterProject/App/PIController.c [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+{
+    return PiCalculation(&PIControllerList.SpeedController, measuredValue);
+}
+
+/*
+ * Functions to check for saturation
+ */
+int isOutputSaturatedPositive(PIobject *PIcontroller)
+{
+    return PIcontroller->previousOutput > PIcontroller->saturationLimit;
```
Comment (murgui) [inline Software/Development/SemesterProject/App/PIController.c]: I think here the variable to use is PIOutput

File: Software/Development/SemesterProject/App/PIController.c [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+int isOutputSaturatedNegative(PIobject * PIcontroller)
+{
+    return PIcontroller->previousOutput < -(PIcontroller->saturationLimit);
```
Comment (murgui) [inline Software/Development/SemesterProject/App/PIController.c]: I think here the variable to use is PIOutput

**Review by murgui — CHANGES_REQUESTED** [review]
Nice looking,

I'm not convinced with the way of setting the reference, because anyway it will be set every time from the controller, then I think it's not worth creating a separate function, but should be an argument of the PI.

Other than that, buen trabajo.

File: Software/Development/SemesterProject/App/PIController.c (hunk 7 above)
Comment (NicolaiFransen) [reply Software/Development/SemesterProject/App/PIController.c]: You are right

File: Software/Development/SemesterProject/App/Include/PIController.h (hunk 6 above)
Comment (NicolaiFransen) [reply Software/Development/SemesterProject/App/Include/PIController.h]: integrationOfError¿?¿

**Review by murgui — APPROVED** [review]
Great job, Ship it!

**Review by AitorTeran — APPROVED** [review]
I dont really understand this but sure...

**Pull request merged**